patterns/strategy.md

```diff
@@ -0,0 +1,153 @@
+# Strategy (aka Policy)
+
+## Description
+
+The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern) is a technique that enables separation of concerns in object-oriented software design.
+It also allows to achieve decoupling software modules known as [Dependency Inversion Principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle).
+The basic idea behind Strategy is given an algorithm, solving a particular problem, we define only skeleton of the algorithm at abstract level, but delegate specific algorithm’s implementation or some of its parts to another class or method.
+A client using algorithm may choose a specific implementation, while the general algorithm workflow remains the same.
+In other words, the abstract specification of the class does not depend on the specific implementation of the derived class, but specific implementation must adhere to the abstract specification.
+This is why we call it Dependency Inversion.
+
+## Motivation
+
+Imagine we are working on a project that generates reports every month. 
+We need the reports to be generated in different formats, e.g., in `JSON` or `Plain Text` formats. 
+But thing very and we don't know what kind of requiremnt we may get in the future. 
+For example, we may need to generate our report in a completly new format, 
+or just modify the existing format. 
+So, in order to make chages to our existing code base as much as possible 
+while keeping our solution adoptable to such changes, 
+we might exploit Strategy design pattern. 
+Rust has powerful trait system which allows to implement this pattern. 
+
+## Example
+
+In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`. 
+`Text` and `Json` are our strategy structs. 
+These strategies have to implement `Formatter` trait.
+
+```rust
+use std::fmt::Write as FmtWrite;
+use std::{error, result};
+type Result = result::Result<String, Box<dyn error::Error>>;
+
+struct Context {
+    pub keys: Vec<String>,
+    pub values: Vec<i32>,
+}
+
+trait Formatter {
+    fn run(&self, context: &Context) -> Result;
+}
+
+struct Report;
+
+impl Report {
+    fn generate<T: Formatter>(g: T) -> Result {
+        //perform here backend operations which should not bother caller...
+        //fetch data from database
+        let keys = vec!["one".to_string(), "two".to_string()];
+        let values = vec![1, 2];
```

Comment:
Shouldn't we show a case with `HashMap` for this? If we use two vecs, it means that it could have different length.

Reply:
Shouldn't length of keys and length of values should be the equal? We assume each value corresponds to each key. I think we it is not that important whether we use `HashMap` or another collection. The main focus should be how we apply Strategy. After all, it is a toy example. I am also planning mention `serde` with use example.

Reply:
After the second thought, using `HashMap` allows getting rid of `Context` and ugly `for` loop on vecs.